add svg animation

tbhartman opened the pull request:
Add SVG animation capabilities (for vector-graphic animations).  I modified the SVG backend so that an XML element tree can be obtained.  The animation is created by looping through the XML element tree for each frame and adding SVG animation elements where needed (animation runs once on load, then on each click over any part of the SVG).  I have tried to use this on several of the animation examples; I know that the dynamic_image script does not work.  I have never made a PR before, so please give feedback.

tacaswell commented:
Cool. 

At a minimum this will need an entry in whats_new (https://github.com/matplotlib/matplotlib/tree/master/doc/users/whats_new).

The travis says you broke all of the svg tests on 2.x so I suspect you have a 3-ism in there someplace.

They all look like

```
======================================================================
FAIL: matplotlib.tests.test_collections.test__EventCollection__set_linestyle.test
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/matplotlib-1.5.dev1-py2.7-linux-x86_64.egg/matplotlib/testing/decorators.py", line 51, in failer
    result = f(*args, **kwargs)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/matplotlib-1.5.dev1-py2.7-linux-x86_64.egg/matplotlib/testing/decorators.py", line 183, in do_test
    figure.savefig(actual_fname, **self._savefig_kwarg)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/matplotlib-1.5.dev1-py2.7-linux-x86_64.egg/matplotlib/figure.py", line 1490, in savefig
    self.canvas.print_figure(*args, **kwargs)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/matplotlib-1.5.dev1-py2.7-linux-x86_64.egg/matplotlib/backend_bases.py", line 2213, in print_figure
    **kwargs)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/matplotlib-1.5.dev1-py2.7-linux-x86_64.egg/matplotlib/backends/backend_svg.py", line 1140, in print_svg
    return self._print_svg(filename, svgwriter, fh_to_close, **kwargs)
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/matplotlib-1.5.dev1-py2.7-linux-x86_64.egg/matplotlib/backends/backend_svg.py", line 1172, in _print_svg
    return renderer.finalize()
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/matplotlib-1.5.dev1-py2.7-linux-x86_64.egg/matplotlib/backends/backend_svg.py", line 218, in finalize
    return self.writer.close()
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/matplotlib-1.5.dev1-py2.7-linux-x86_64.egg/matplotlib/backends/backend_svg.py", line 130, in close
    tree = XMLBuilder.close(self)
  File "/opt/python/2.7.9/lib/python2.7/xml/etree/ElementTree.py", line 1384, in close
    assert len(self._elem) == 0, "missing end tags"
AssertionError: missing end tags
```

This is something that has been at the back of my mind as something we should have for a while now, thanks for doing this!

tacaswell commented on lib/matplotlib/backends/backend_svg.py:
```diff
@@ -1195,19 +1132,23 @@ def print_svg(self, filename, *args, **kwargs):
     def print_svgz(self, filename, *args, **kwargs):
         if is_string_like(filename):
-            fh_to_close = gzipwriter = gzip.GzipFile(filename, 'w')
+            gzipwriter = gzip.GzipFile(filename, 'w')
```
There was something funny recently with how exactly these zip files got closed....

WeatherGod commented:
attn: @dopplershift

tbhartman commented:
@tacaswell I rebased my commit off of master last night.  Fixed the Python2 issue within that commit, and added a whats_new file (also added to the CHANGELOG, is that appropriate?).  Fixed PathEffectRenderer not calling the correct close_group method (so SVG groups were not being closed correctly) and put that in a new commit.  I think that needs fixed regardless of the SVG animation, so I didn't squash those, do I need to?

Travis is now happy!

tacaswell commented on lib/matplotlib/animation.py:
```diff
@@ -559,6 +563,458 @@ def _args(self):
+class _SvgAnimate(object):
+    """an object to carry svg animation information.
+
+    I store the element's attributes for each from of the animation.
```
Can you remove the first-person in the doc strings?

tbhartman replied:
Done and extended doctring for that class.

tacaswell commented on lib/matplotlib/animation.py:
```diff
@@ -559,6 +563,458 @@ def _args(self):
+
+    @classmethod
+    def populate(cls, element, frame, under_defs=False):
+        """add an _SvgAnimate object to an XML *element*'s attributes.
```
Can you format this as a numpydoc string?

tbhartman replied:
Done.

tacaswell commented:
@mdboom Who understands the svg backend and can review this?

tacaswell commented:
@tbhartman This needs a rebase (probably due to the recent salt/no salt changes to the svg backend).

I have re-milestoned this as 'proposed next point release'.  If this is ready in time it can go in, but it is not a blocker for the next release.

dopplershift commented:
Animation stuff looks sound; I can't really speak to the SVG stuff.

tbhartman commented:
Great.  I will not be able to do this for at least a week.  I will update
you when I rebase.
On Jun 16, 2015 3:04 AM, "Thomas A Caswell" <notifications@github.com>
wrote:

> @tbhartman <https://github.com/tbhartman> This needs a rebase (probably
> due to the recent salt/no salt changes to the svg backend).
>
> I have re-milestoned this as 'proposed next point release'. If this is
> ready in time it can go in, but it is not a blocker for the next release.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/4255#issuecomment-112259381>
> .
>

WeatherGod commented:
ping @tbhartman

tbhartman commented:
I have seen this but have not had access to my computer which contains all of this.  When I do, I will update.

OceanWolf commented:
I have only taken a cursory glance at this, but I do note that ``XMLWriter`` has had quite a few changes with some methods removed, how does this affect Backwards Compatibility?

tacaswell commented on lib/matplotlib/backends/backend_svg.py:
```diff
@@ -26,196 +26,120 @@
 backend_version = __version__
 
-# ----------------------------------------------------------------------
-# SimpleXMLWriter class
```
Please leave this license unless you really have a 100% clean-room re-write (which I do not think this is as there is one method which is identical).  I err on the side of caution in these matters.

OceanWolf replied:
Just one question, why don't we use python's inbuilt XML library?  It has been part of the library since python 1.5 (so I have just read)

tbhartman replied:
@OceanWolf I may have not looked into this as much as I should have.  I wanted to keep the API of the old XMLWriter class (at least to the extent that it is used within MPL).  I believe the simplest method to do this is by extending the xml.etree.TreeBuilder class.  I also wanted to be sure that the output XML was formatted the same so that all the tests of XML results all passed, so the XMLWriter class was still necessary.

tbhartman replied:
@tacaswell I'll need to go back and compare the old code but I think you are right, that I kept the end method and made it into the xml_write function.  Do I place the license just before that function then?

mdboom replied:
The reason we use this rather than the Python library's built-in TreeBuilder is due to streaming.  The changes here build an XML tree in memory, and then write it out.  That, at least when I originally wrote this, was much slower and less memory efficient, particularly for large paths.  There's a very good reason this XMLWriter class was used, and I think "because Python has a tree builder in the standard library" is not a good enough reason to completely change the approach here.

tbhartman replied:
@mdboom I'm trying to look at this again.  My thought is to modify the XMLWriter class with an argument on the __init__ method to give an option to write out the tree immediately rather than build a tree in memory.  I can then write out images without building the tree, but if an animation is being made, build the tree in memory.

mdboom replied:
That seems like a reasonable middle round.  I look forward to seeing it...  Thanks!

tacaswell commented:
@OceanWolf I think if someone is relying on mpl for an XML writer that may be their problem.

OceanWolf commented:
@tacaswell Probably, but it doesn't hurt to mention these things, tis a public class after all ;).

dopplershift commented:
I'm not sure about how MPL is handling things, but public class does not imply public API in a language like Python.

tbhartman commented:
I have been traveling internationally for several weeks but am back and will take a look at this soon and push an update.

WeatherGod commented:
needs rebase @tbhartman

sughodke commented:
Its too bad this feature was left on the table. If folks still care about it, I would advocate to start up these discussions again.  @WeatherGod  @tacaswell

tacaswell commented:
@sughodke Any help on this would be appreciated!

dbrnz commented:
This looks as if this is what I'm wanting so am currently rebasing this PR; I'll create a new PR when done.

jklymak commented:
Closing in lieu of #9086....  Thanks for starting @tbhartman !
